backup: add simpler RESTORE span computation

dt opened the pull request:
Previously we used `overlappingCoveringMerge` during `RESTORE` to do two things:
a) compute a covering that partitioned the restoring spans and assigned files to be restored to those
partitions, and
b) verify that the key spans being restored were covered over all time spans by the backups.

The former produced a much more granularly partitioned covering than was
required, splitting the overall covering at all file boundaries, in all layers.
When the spans of files in incremental layers did not line up with the spans in other 
layers, this produced a huge number of tiny partitions, with each file having to
be copied to many, many partitions it overlapped.

Instead, this change introduces a new method of deriving this covering. It now
uses the files of the first backup that cover any part of the key span being restored
to define partitions and then assigns later files to any partitions they overlap. This 
produces only one partition per file that increases the covered spans, so if all the
inc backups are of the same tables/spans, it'd just be file spans of the first backup.

This simpler approach no longer has the side effect of also validating that all the
key spans are covered at all time spans that the overlappingCoveringMerge did,
so when used the caller needs to also perform their own coverage verification check,
which can use the cheaper method previously introduced in #74393.


There are more extensive benchmark results the last commit in this PR, but for a quick summary:
```
name                                                         old time/op    new time/op    delta
RestoreEntryCover/numBackups=1/numFiles=10000/numSpans=100     48.5ms ± 2%    26.3ms ± 1%   -45.68%  (p=0.008 n=5+5)
RestoreEntryCover/numBackups=24/numFiles=10000/numSpans=100     1.26s ± 1%     0.33s ± 1%   -74.21%  (p=0.008 n=5+5)
RestoreEntryCover/numBackups=96/numFiles=10000/numSpans=100     9.80s ± 1%     1.47s ± 1%   -84.99%  (p=0.036 n=3+5)

name                                                         old alloc/op   new alloc/op   delta
RestoreEntryCover/numBackups=1/numFiles=10000/numSpans=100     44.9MB ± 0%     8.9MB ± 0%   -80.21%  (p=0.008 n=5+5)
RestoreEntryCover/numBackups=24/numFiles=10000/numSpans=100     917MB ± 0%      43MB ± 0%   -95.31%  (p=0.008 n=5+5)
RestoreEntryCover/numBackups=96/numFiles=10000/numSpans=100    4.93GB ± 0%    0.24GB ± 0%   -95.06%  (p=0.000 n=3+4)
```

Release note (performance improvement): The merging of incremental backup layers during RESTORE now uses a simpler and less memory intensive algorithm.

dt committed:
backup: move makeImportSpans to own file

Release note: none.

dt committed:
backup: add simpler RESTORE span computation

Previously we used overlappingCoveringMerge during RESTORE to a) compute
a covering of the restoring spans and assign files to be restored to the
partitions of it while b) verifying that the key spans being restored were
covered over all time spans by the backups.

The former produced a much more granularly partitioned covering than was
required, splitting the overall covering at all file boundaries in all layers,
resulting in many tiny partitions and individual files having to be copied
to many, many partitions they overlap. Instead, this change introduces a new
method of deriving this covering which is to more or less just use the files
of the first backup that covers some part of the key span to define partitions
and then assigning later files to any partition they overlap. This
simpler approach no longer has the side effect of also validating that all the
key spans are covered at all time spans, so when used it needs to also call the
standalone coverage verification check previously introduced.

Release note (performance improvement): The merging of incremental backup layers during RESTORE now uses a simpler and less memory intensive algorithm.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/74394)
<!-- Reviewable:end -->

dt commented:
First commit is #74394. Need to write some tests and then benchmark it too but wanted to get this out for initial look.

dt committed:
backupccl: extract mock chain helper from benchmark

Release note: none.

dt committed:
backupccl: benchmark restore span covering methods

```
name                                                         old time/op    new time/op    delta
RestoreEntryCover/numBackups=1/numFiles=0/numSpans=100          447µs ± 2%     490µs ± 2%    +9.39%  (p=0.008 n=5+5)
RestoreEntryCover/numBackups=1/numFiles=100/numSpans=100        823µs ± 1%     752µs ± 1%    -8.63%  (p=0.008 n=5+5)
RestoreEntryCover/numBackups=1/numFiles=10000/numSpans=100     48.5ms ± 2%    26.3ms ± 1%   -45.68%  (p=0.008 n=5+5)
RestoreEntryCover/numBackups=2/numFiles=0/numSpans=100          650µs ± 1%     661µs ± 1%    +1.66%  (p=0.032 n=5+5)
RestoreEntryCover/numBackups=2/numFiles=100/numSpans=100       1.24ms ± 1%    1.04ms ± 3%   -15.79%  (p=0.008 n=5+5)
RestoreEntryCover/numBackups=2/numFiles=10000/numSpans=100     75.7ms ± 1%    38.4ms ± 0%   -49.32%  (p=0.008 n=5+5)
RestoreEntryCover/numBackups=24/numFiles=0/numSpans=100        4.78ms ± 0%    5.42ms ± 1%   +13.32%  (p=0.008 n=5+5)
RestoreEntryCover/numBackups=24/numFiles=100/numSpans=100      16.2ms ± 1%     8.9ms ± 1%   -45.45%  (p=0.008 n=5+5)
RestoreEntryCover/numBackups=24/numFiles=10000/numSpans=100     1.26s ± 1%     0.33s ± 1%   -74.21%  (p=0.008 n=5+5)
RestoreEntryCover/numBackups=96/numFiles=0/numSpans=100        18.3ms ± 0%    24.7ms ± 0%   +34.85%  (p=0.008 n=5+5)
RestoreEntryCover/numBackups=96/numFiles=100/numSpans=100       117ms ± 1%      40ms ± 1%   -65.87%  (p=0.008 n=5+5)
RestoreEntryCover/numBackups=96/numFiles=10000/numSpans=100     9.80s ± 1%     1.47s ± 1%   -84.99%  (p=0.036 n=3+5)

name                                                         old coverSize  new coverSize  delta
RestoreEntryCover/numBackups=1/numFiles=0/numSpans=100           0.00           0.00           ~     (all equal)
RestoreEntryCover/numBackups=1/numFiles=100/numSpans=100          100 ± 0%       100 ± 0%      ~     (all equal)
RestoreEntryCover/numBackups=1/numFiles=10000/numSpans=100      17.2k ± 0%      5.8k ± 0%   -66.39%  (p=0.008 n=5+5)
RestoreEntryCover/numBackups=2/numFiles=0/numSpans=100           0.00           0.00           ~     (all equal)
RestoreEntryCover/numBackups=2/numFiles=100/numSpans=100          151 ± 0%       100 ± 0%   -33.77%  (p=0.008 n=5+5)
RestoreEntryCover/numBackups=2/numFiles=10000/numSpans=100      25.7k ± 0%      5.7k ± 0%   -77.66%  (p=0.008 n=5+5)
RestoreEntryCover/numBackups=24/numFiles=0/numSpans=100          0.00           0.00           ~     (all equal)
RestoreEntryCover/numBackups=24/numFiles=100/numSpans=100       1.64k ± 0%     0.10k ± 0%   -93.83%  (p=0.008 n=5+5)
RestoreEntryCover/numBackups=24/numFiles=10000/numSpans=100      150k ± 0%        6k ± 0%   -96.32%  (p=0.008 n=5+5)
RestoreEntryCover/numBackups=96/numFiles=0/numSpans=100          0.00           0.00           ~     (all equal)
RestoreEntryCover/numBackups=96/numFiles=100/numSpans=100       5.04k ± 0%     0.13k ± 0%   -97.42%  (p=0.008 n=5+5)
RestoreEntryCover/numBackups=96/numFiles=10000/numSpans=100      490k ± 0%        5k ± 0%      ~     (p=0.357 n=3+5)

name                                                         old alloc/op   new alloc/op   delta
RestoreEntryCover/numBackups=1/numFiles=0/numSpans=100          180kB ± 0%      98kB ± 0%   -45.67%  (p=0.008 n=5+5)
RestoreEntryCover/numBackups=1/numFiles=100/numSpans=100        410kB ± 0%     134kB ± 0%   -67.35%  (p=0.008 n=5+5)
RestoreEntryCover/numBackups=1/numFiles=10000/numSpans=100     44.9MB ± 0%     8.9MB ± 0%   -80.21%  (p=0.008 n=5+5)
RestoreEntryCover/numBackups=2/numFiles=0/numSpans=100          258kB ± 0%     193kB ± 0%   -25.00%  (p=0.008 n=5+5)
RestoreEntryCover/numBackups=2/numFiles=100/numSpans=100        619kB ± 0%     244kB ± 0%   -60.65%  (p=0.008 n=5+5)
RestoreEntryCover/numBackups=2/numFiles=10000/numSpans=100     68.8MB ± 0%    10.8MB ± 0%   -84.28%  (p=0.008 n=5+5)
RestoreEntryCover/numBackups=24/numFiles=0/numSpans=100        1.80MB ± 0%    2.25MB ± 0%   +24.56%  (p=0.008 n=5+5)
RestoreEntryCover/numBackups=24/numFiles=100/numSpans=100      8.66MB ± 0%    2.70MB ± 0%   -68.84%  (p=0.008 n=5+5)
RestoreEntryCover/numBackups=24/numFiles=10000/numSpans=100     917MB ± 0%      43MB ± 0%   -95.31%  (p=0.008 n=5+5)
RestoreEntryCover/numBackups=96/numFiles=0/numSpans=100        6.89MB ± 0%    8.32MB ± 0%   +20.71%  (p=0.008 n=5+5)
RestoreEntryCover/numBackups=96/numFiles=100/numSpans=100      54.8MB ± 0%    10.0MB ± 0%   -81.80%  (p=0.008 n=5+5)
RestoreEntryCover/numBackups=96/numFiles=10000/numSpans=100    4.93GB ± 0%    0.24GB ± 0%   -95.06%  (p=0.000 n=3+4)
```

Release note: none.

miretskiy commented on pkg/ccl/backupccl/restore_span_covering.go:
```diff
@@ -246,3 +248,75 @@ rangeLoop:
+// makeSimpleImportSpans partitions the spans of requiredSpans into a covering
+// of RestoreSpanEntry's which each have all overlapping files from the passed
+// backups assigned to them. The spans of requiredSpans are trimmed/removed
+// based on the lowWaterMark.
```
perhaps indicate that backups must also be ordered (incremental chain).

miretskiy commented on pkg/ccl/backupccl/restore_span_covering.go:
```diff
@@ -247,3 +249,75 @@ rangeLoop:
+// backups assigned to them. The spans of requiredSpans are trimmed/removed
+// based on the lowWaterMark.
+func makeSimpleImportSpans(
+	requiredSpans roachpb.Spans,
```
perhaps []roachpb.Span instead?

miretskiy commented on pkg/ccl/backupccl/restore_span_covering.go:
```diff
@@ -247,3 +249,75 @@ rangeLoop:
+					if len(cover) == spanCoverStart {
+						cover = append(cover, makeEntry(span.Key, sp.EndKey, fileSpec))
+					} else {
+						// Add each file to every matching partition in the cover.
```
do you think this code could benefit from a ascii-art comment -- like the one on `frontier.SpanEntries`?

dt replied:
ugh, yes.

I started one when I first wrote this and then removed it since I was spending more time trying to fix it up than the code it felt like, but I guess a little more investment now will save re-figuring it out later.

Added.

dt commented on pkg/ccl/backupccl/restore_job.go:
```diff
@@ -418,11 +418,23 @@ func restore(
-	if err != nil {
-		return emptyRowCount, errors.Wrapf(err, "making import requests for %d backups", len(backupManifests))
+
+	const useSimpleImportSpans = true
```
No, if checkCoverage fails we return the error, right? We never call makeImportSpans without a code change

miretskiy commented on pkg/ccl/backupccl/restore_span_covering.go:
```diff
@@ -247,3 +249,93 @@ rangeLoop:
+//  [l, m): 9
+// This example is tested in TestRestoreEntryCoverExample.
+func makeSimpleImportSpans(
+	requiredSpans []roachpb.Span,
```
Well, I prefer direct types unless I need those extra things provided by an alias.  But I agree, it's a nit.

miretskiy reviewed (approved):
This looks great to me.  Perhaps other people on bulk should also look it over.

One thing (re todo on bsearch) you could try doing is to implement interval.Interface via backup file so that you can construct an interval tree of backup files.

adityamaru reviewed (approved):
LGTM!! I'm hopeful this is the panacea for our memory woes.

dt commented:
TFTRs!

bors r+

dt commented on pkg/ccl/backupccl/restore_span_covering.go:
```diff
@@ -247,3 +249,93 @@ rangeLoop:
+								cover[i].Files = append(cover[i].Files, fileSpec)
+							}
+							// Later files start later, so this cover ends before this file,
+							// it ends before them too and they can start searching after it.
```
Ah, good catch, should say "if". I'll get it in followup rather than eat another CI now.

dt commented:
blathers backport 21.2
